src/sage/combinat/integer_vector.py

```diff
@@ -900,6 +900,63 @@ def __contains__(self, x):
             return False
         return sum(x) == self.n
 
+    def rank(self, x):
+        """
+        Return the rank of a given element.
+
+        INPUT:
+
+        - ``x`` -- a list with ``sum(x) == n``
+
+        TESTS::
+
+            sage: IntegerVectors(k=5).rank([0,0,1,2,2])
+            243
+        """
+        if sum(x)!=self.n:
+            raise ValueError("argument is not a member of IntegerVectors({},{})".format(self.n, None))
+            
+        n, k, r = self.n, len(x), 0
+        for i in range(k):
+            r += binomial(i+n-1, n)
```

Comment:
While we are on the topic of efficiency, it can be useful to notice that
```sum(binomial(m, k) for m in range(n)) == binomial(n, k+1)```
see Knuth *The art of computer programming* §1.2.6E or [Wikipedia](https://en.wikipedia.org/wiki/Binomial_coefficient#Sums_of_the_binomial_coefficients).

The right hand site can be slightly faster (try $n=10^7$ and $k=17$).